<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from active_caster.xacro            | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="dyno_active_caster" xmlns:xacro="http://ros.org/wiki/xacro">
  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="dark">
    <color rgba="0.3 0.3 0.3 1.0"/>
  </material>
  <material name="light_black">
    <color rgba="0.4 0.4 0.4 1.0"/>
  </material>
  <material name="blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <material name="light_blue">
    <color rgba="0.3 0.3 0.8 1.0"/>
  </material>
  <material name="green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>
  <material name="grey">
    <color rgba="0.5 0.5 0.5 1.0"/>
  </material>
  <material name="orange">
    <color rgba="1.0 0.423529411765 0.0392156862745 1.0"/>
  </material>
  <material name="brown">
    <color rgba="0.870588235294 0.811764705882 0.764705882353 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
    </plugin>
    <plugin filename="libgazebo_ros_p3d.so" name="ground_truth">
      <frameName>map</frameName>
      <bodyName>base_footprint</bodyName>
      <topicName>base_pose_ground_truth</topicName>
      <updateRate>30.0</updateRate>
    </plugin>
  </gazebo>
  <link name="base_link">
    <visual>
      <origin rpy="0 1.57079632679 0" xyz="0.2 0 0"/>
      <geometry>
        <cylinder length="0.02" radius="0.01"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.2"/>
      <geometry>
        <cylinder length="0.1" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="10.0"/>
      <inertia ixx="0.0833333333333" ixy="0" ixz="0" iyy="0.0833333333333" iyz="0" izz="0.15"/>
    </inertial>
  </link>
  <!-- Base footprint -->
  <link name="base_footprint"/>
  <joint name="base_footprint_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.055"/>
    <parent link="base_footprint"/>
    <child link="base_link"/>
  </joint>
  <!-- Macros -->
  <!-- Triangle -->
  <!--
    <active_caster name="fl_caster" parent="base_link" x_offset="${wheel_separation/(sqrt(3)*2)}" y_offset="${wheel_separation/2}"/>
    <active_caster name="fr_caster" parent="base_link" x_offset="${wheel_separation/(sqrt(3)*2)}" y_offset="${-wheel_separation/2}"/>
    <active_caster name="rc_caster" parent="base_link" x_offset="${-wheel_separation/sqrt(3)}" y_offset="0"/>
    -->
  <!-- Mount -->
  <link name="fl_caster_mount_link">
    <!--
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder length="0.001" radius="0.001"/>
        </geometry>
      </visual>
       -->
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="fl_caster_mount_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.165 0.165 0"/>
    <parent link="base_link"/>
    <child link="fl_caster_mount_link"/>
  </joint>
  <!-- Steer -->
  <link name="fl_caster_steer_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.03"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="fl_caster_steer_joint" type="continuous">
    <parent link="fl_caster_mount_link"/>
    <child link="fl_caster_steer_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Wheel -->
  <link name="fl_caster_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.5"/>
      <inertia ixx="5.9375e-05" ixy="0" ixz="0" iyy="5.9375e-05" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <joint name="fl_caster_wheel_joint" type="continuous">
    <origin rpy="1.57079632679 0 0" xyz="-0.04 0 -0.034"/>
    <parent link="fl_caster_steer_link"/>
    <child link="fl_caster_wheel_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Transmission -->
  <transmission name="fl_caster_steer_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="fl_caster_steer_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="fl_caster_steer_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="fl_caster_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="fl_caster_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="fl_caster_wheel_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <!-- Contact model for the wheel surface -->
  <gazebo reference="fl_caster_wheel_link">
    <!--
     <kp>0.1</kp>
     <kd>0.1</kd>
      -->
    <kp>100000.0</kp>
    <kd>100.0</kd>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <fdir1>1 0 0</fdir1>
    <maxVel>1.0</maxVel>
    <minDepth>0.001</minDepth>
    <implicitSpringDamper>1</implicitSpringDamper>
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <!-- Mount -->
  <link name="fr_caster_mount_link">
    <!--
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder length="0.001" radius="0.001"/>
        </geometry>
      </visual>
       -->
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="fr_caster_mount_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.165 -0.165 0"/>
    <parent link="base_link"/>
    <child link="fr_caster_mount_link"/>
  </joint>
  <!-- Steer -->
  <link name="fr_caster_steer_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.03"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="fr_caster_steer_joint" type="continuous">
    <parent link="fr_caster_mount_link"/>
    <child link="fr_caster_steer_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Wheel -->
  <link name="fr_caster_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.5"/>
      <inertia ixx="5.9375e-05" ixy="0" ixz="0" iyy="5.9375e-05" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <joint name="fr_caster_wheel_joint" type="continuous">
    <origin rpy="1.57079632679 0 0" xyz="-0.04 0 -0.034"/>
    <parent link="fr_caster_steer_link"/>
    <child link="fr_caster_wheel_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Transmission -->
  <transmission name="fr_caster_steer_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="fr_caster_steer_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="fr_caster_steer_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="fr_caster_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="fr_caster_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="fr_caster_wheel_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <!-- Contact model for the wheel surface -->
  <gazebo reference="fr_caster_wheel_link">
    <!--
     <kp>0.1</kp>
     <kd>0.1</kd>
      -->
    <kp>100000.0</kp>
    <kd>100.0</kd>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <fdir1>1 0 0</fdir1>
    <maxVel>1.0</maxVel>
    <minDepth>0.001</minDepth>
    <implicitSpringDamper>1</implicitSpringDamper>
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <!-- Mount -->
  <link name="rl_caster_mount_link">
    <!--
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder length="0.001" radius="0.001"/>
        </geometry>
      </visual>
       -->
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="rl_caster_mount_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-0.165 0.165 0"/>
    <parent link="base_link"/>
    <child link="rl_caster_mount_link"/>
  </joint>
  <!-- Steer -->
  <link name="rl_caster_steer_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.03"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="rl_caster_steer_joint" type="continuous">
    <parent link="rl_caster_mount_link"/>
    <child link="rl_caster_steer_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Wheel -->
  <link name="rl_caster_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.5"/>
      <inertia ixx="5.9375e-05" ixy="0" ixz="0" iyy="5.9375e-05" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <joint name="rl_caster_wheel_joint" type="continuous">
    <origin rpy="1.57079632679 0 0" xyz="-0.04 0 -0.034"/>
    <parent link="rl_caster_steer_link"/>
    <child link="rl_caster_wheel_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Transmission -->
  <transmission name="rl_caster_steer_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rl_caster_steer_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rl_caster_steer_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="rl_caster_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rl_caster_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rl_caster_wheel_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <!-- Contact model for the wheel surface -->
  <gazebo reference="rl_caster_wheel_link">
    <!--
     <kp>0.1</kp>
     <kd>0.1</kd>
      -->
    <kp>100000.0</kp>
    <kd>100.0</kd>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <fdir1>1 0 0</fdir1>
    <maxVel>1.0</maxVel>
    <minDepth>0.001</minDepth>
    <implicitSpringDamper>1</implicitSpringDamper>
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <!-- Mount -->
  <link name="rr_caster_mount_link">
    <!--
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder length="0.001" radius="0.001"/>
        </geometry>
      </visual>
       -->
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="rr_caster_mount_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-0.165 -0.165 0"/>
    <parent link="base_link"/>
    <child link="rr_caster_mount_link"/>
  </joint>
  <!-- Steer -->
  <link name="rr_caster_steer_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.03"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.01" radius="0.01"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.000166666666667" ixy="0" ixz="0" iyy="0.000166666666667" iyz="0" izz="0.000166666666667"/>
    </inertial>
  </link>
  <joint name="rr_caster_steer_joint" type="continuous">
    <parent link="rr_caster_mount_link"/>
    <child link="rr_caster_steer_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Wheel -->
  <link name="rr_caster_wheel_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.015" radius="0.02"/>
      </geometry>
      <material name="orange"/>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.5"/>
      <inertia ixx="5.9375e-05" ixy="0" ixz="0" iyy="5.9375e-05" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <joint name="rr_caster_wheel_joint" type="continuous">
    <origin rpy="1.57079632679 0 0" xyz="-0.04 0 -0.034"/>
    <parent link="rr_caster_steer_link"/>
    <child link="rr_caster_wheel_link"/>
    <axis xyz="0 0 1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <!-- Transmission -->
  <transmission name="rr_caster_steer_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rr_caster_steer_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rr_caster_steer_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="rr_caster_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rr_caster_wheel_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="rr_caster_wheel_joint_motor">
      <mechanicalReduction>10</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <!-- Contact model for the wheel surface -->
  <gazebo reference="rr_caster_wheel_link">
    <!--
     <kp>0.1</kp>
     <kd>0.1</kd>
      -->
    <kp>100000.0</kp>
    <kd>100.0</kd>
    <mu1>1.0</mu1>
    <mu2>1.0</mu2>
    <fdir1>1 0 0</fdir1>
    <maxVel>1.0</maxVel>
    <minDepth>0.001</minDepth>
    <implicitSpringDamper>1</implicitSpringDamper>
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <gazebo reference="base_link">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="base_footprint">
    </gazebo>
</robot>
